functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,137 @@
+import torch
+from torch._C import DispatchKey, DispatchKeySet, ExcludeDispatchKeyGuard
+from functorch.experimental.ops import PyOperator
+from torch.utils._pytree import tree_flatten
+from torch.fx.experimental.proxy_tensor import get_isolated_graphmodule, disable_proxy_modes_tracing
+import torch.utils._pytree as pytree
+from torch.dispatch.dispatcher import dispatcher_singleton
+
+"""
+We're going to define a `cond` operation.
+In order to do this, we need implementations for each of the dispatch keys.
+"""
+from contextlib import contextmanager
+
+# TODO(voz): Move out somewhere else once other py dispatched ops need it
+@contextmanager
+def suspend_mode(mode):
+    assert(mode is not None), "Cannot suspend None mode"
+    torch._C._set_torch_dispatch_mode(None)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(mode)
+
+@contextmanager
+def enable_mode(mode):
+    curr_mode = torch._C._get_torch_dispatch_mode()
+    torch._C._set_torch_dispatch_mode(mode)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(curr_mode)
+
+
+def trace_cond(proxy_mode, func_overload, args, kwargs=None):
+    assert kwargs is None or not kwargs
+    pred, true_fn, false_fn, operands = args
+
+    def _unwrap_proxy(e):
+        if hasattr(e, "__dict__") and proxy_mode.tracer in e.__dict__:
+            proxy_out = e.__dict__[proxy_mode.tracer]
+            return proxy_out.proxy
+        return e
+
+
+    if isinstance(operands, torch.Tensor):
+        operands = [operands]  # Little hack because * on a single ProxyTensor unpacks it
+    else:
+        operands = operands
+
+    true_graph = get_isolated_graphmodule(true_fn, operands, {})
+    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+    true_name = "true_graph"
+    false_name = "false_graph"
+    proxy_mode.tracer.root.register_module(true_name, true_graph)
+    proxy_mode.tracer.root.register_module(false_name, false_graph)
+
+    true_args = []
+    for node in true_graph.graph.nodes:
+        if node.op == 'output':
+            true_args.extend(*node.args)
+
+    false_args = []
+    for node in false_graph.graph.nodes:
+        if node.op == 'output':
+            false_args.extend(*node.args)
+
+    assert(len(true_args) == len(false_args))
+    for i in range(len(true_args)):
+        t_arg_meta = true_args[i]
+        f_arg_meta = false_args[i]
+        # WIP don't look at this yet
+        # print(t_arg_meta.meta)
+        # print(f_arg_meta.meta)
+        # assert(t_arg_meta.meta == f_arg_meta.meta)
```

Comment:
yes please lol

Reply:
working on it ;) the op outvars are still missing meta